relay-server/src/services/buffer/envelope_buffer/mod.rs

```diff
@@ -250,13 +280,73 @@ where
         changed
     }
 
+    /// Evicts the least recently used stacks.
+    #[allow(dead_code)]
+    pub async fn evict(&mut self) {
+        #[derive(Debug, Copy, Clone)]
+        struct LRUItem(StackKey, Readiness, Instant);
+
+        impl PartialEq for LRUItem {
+            fn eq(&self, other: &Self) -> bool {
+                self.0 == other.0
+            }
+        }
+
+        impl PartialOrd for LRUItem {
+            fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+                Some(self.cmp(other))
+            }
+        }
+
+        impl Eq for LRUItem {}
+
+        impl Ord for LRUItem {
+            fn cmp(&self, other: &Self) -> Ordering {
+                match (self.1.ready(), other.1.ready()) {
+                    (true, false) => Ordering::Greater,
+                    (false, true) => Ordering::Less,
+                    _ => self.2.cmp(&other.2),
+                }
+            }
+        }
+
+        let mut lru: BinaryHeap<LRUItem> = BinaryHeap::new();
+        for (queue_item, priority) in self.priority_queue.iter() {
+            let lru_item = LRUItem(queue_item.key, priority.readiness, queue_item.last_update);
+
+            // If we exceed the size, we want to pop the greatest element only if we have a smaller
+            // element, so that we end up with the smallest elements which are the ones with the
+            // lowest priority.
+            if lru.len() >= self.max_evictable_stacks {
+                let Some(top_lru_item) = lru.peek() else {
+                    continue;
+                };
+
+                if lru_item < *top_lru_item {
+                    lru.pop();
+                }
+            }
+
+            lru.push(lru_item);
+        }
+
+        // We go over each element and remove it from the stack. After removal, we will evict
+        // elements from each popped stack.
+        for lru_item in lru {
+            if let Some(mut stack) = self.pop_stack(lru_item.0) {
+                stack.evict().await;
```

Comment:
If we are using disk, it's not the most efficient thing to run a query for each project key pair but we can investigate the performance of this later on.